keps/sig-cluster-lifecycle/addons/20191013-install-addons-via-kubeadm.md

```diff
@@ -0,0 +1,450 @@
+# this requires adding a feature-gates flag to this command
+```
+TODO: Discuss this alternative:
+```shell
+kubeadm config print addon-defaults
+```
+Similar U/X can be expected for `kubeadm config migrate` and `kubeadm config view`
+
+The user may extend the default kubeadm config using the `AddonInstallerConfiguration` kind.
+The API details are still being finalized. Usage may look like this:
+```yaml
+---
+apiVersion: addons.config.k8s.io/v1alpha1
+kind: AddonInstallerConfiguration
+addons:
+- name: kube-proxy
+  ref: github.com/kubernetes/kubernetes//cluster/addons/kustomize/kube-proxy/?ref=v1.17.0
+- name: coredns-operator
+  ref: github.com/kubernetes-sigs/addon-operators//kustomize/coredns-operator/?ref=v0.1.0
+- name: weavenet
+  ref: oci+kustomize://weaveworks/kustomize-weavenet:v2.5.2
+    # kustomize OCI backend currently unimplemented
+    # POC packaging:  https://github.com/ecordell/kpg
+- name: my-addon1
+  ref: ../my-local-dir/addon1/
+```
+
+It may be in scope for kubeadm to internally patch the `AddonInstallerConfiguration` with the contents of user-supplied ComponentConfigs such as `KubeProxyConfiguration`.
+Various patchList fields may need to be added to the `Addon` type of the library's API to support this specific U/X as well as more lightweight user-extension of addons.
+
+The AddonInstaller feature composes with kubeadm's global `--dry-run` flag:
+- on `init`, the intended addons (and perhaps manifests) can simply be printed.
+- on `upgrade diff` and `upgrade apply`, the expected difference relative to the cluster can be output.
+  `kubectl` already does this currently.
+
+The difference in behavior between a dryRun init and upgrade is caused by the presence of a working KUBECONFIG and control-plane.
+
+Air-gap support may be accomplished through the mirroring of git repositories or images containing the kustomize dir.
+Mirroring git repositories, cloning git repositories to a local filesystem, and hosting OCI images would seem to all be a reasonable capabilities for a user in an air-gapped environment. The API allows a user to reference addon sources for all of these use-cases.
+
+#### Links
+
+@fabriziopandini began a design doc for kubeadm integration:
+- https://docs.google.com/document/d/1FZA873LBXf-aG4UULOk96MhT8xm_nPHq5Vu0z61S8pk
+
+More generic User Stories have been collected by @dholbach:
+- doc: https://docs.google.com/document/d/17NH4xcFdeh4NVcIBjMQNK2P27j9xrsm4sCFTXdZqopg
+- sheet: https://docs.google.com/spreadsheets/d/1Np0aOQYyiRqRQYOUoMcUwP_2Tt9b39HUwJXQIgW9rYA
+
+### Implementation Details/Notes/Constraints
+
+We intend for the installer library and API to live in `sigs.k8s.io/addon-operators/installer`.
+Core kustomize addons and images may be hosted in a number of git repositories or image registries.
+
+Internally, the addon-installer library execs `kubectl -k`.
+This introduces a binary and PATH dependency for `kubeadm` when using this feature.
```

Comment:
Is this a dependency that applies only to control plane nodes?

Reply:
Yes, it's only required for nodes expected to perform `init`/`upgrade`.